app/src/main/java/mozilla/lockbox/LockboxAutofillService.kt

```diff
@@ -121,27 +121,88 @@ class LockboxAutofillService(
+                    is AutofillAction.Authenticate -> builder.buildAuthenticationFillResponse(this)
+                    is AutofillAction.Cancel -> null
                     is AutofillAction.Error -> {
                         callback.onFailure(getString(R.string.autofill_error_toast, appName, it.error.localizedMessage))
                         null
                     }
                 }.asOptional()
             }
             .filterNotNull()
+            .doOnComplete {
+                compositeDisposable.clear()
+            }
             .subscribe({
-                callback.onSuccess(it.value)
+                callback.onSuccess(it)
             }, {
                 log.error(throwable = it)
             })
             .addTo(compositeDisposable)
     }
 
-    // to be implemented in issue #217
-    override fun onSaveRequest(request: SaveRequest, callback: SaveCallback) {}
+    private fun intializeService() {
+        isRunning = true
+
+        val contextInjectables = arrayOf(
+            telemetryStore,
+            securePreferences,
+            accountStore,
+            fxaSupport
+        )
+
+        contextInjectables.forEach {
+            it.injectContext(this)
+        }
+        dispatcher.dispatch(LifecycleAction.AutofillStart)
+    }
+
+    override fun onSaveRequest(request: SaveRequest, callback: SaveCallback) {
+        val parsedStructure = request.clientState?.let {
+            it.classLoader = ParsedStructure::class.java.classLoader
+            it.getParcelable<ParsedStructure>(Constant.Key.parsedStructure)
+        } ?: return callback.onFailure("Bundle missing")
+
+        val structure = request.fillContexts.last().structure
+        val nodeNavigator = ViewNodeNavigator(structure, parsedStructure.packageName)
+
+        val autofillItem = AutofillTextValueBuilder(parsedStructure, nodeNavigator).build()
+
+        // The SaveInfo we have fillled in means that we'll only get here if there's a username and password.
+        // We can safely bail knowing we'll never need to here.
+        val capturedUsername = autofillItem.username
+        val capturedPassword = autofillItem.password ?: return callback.onFailure("Password missing")
+
+        // According to the AsyncLoginsStorage docs:
+        // "If login has an empty id field, then a GUID will be generated automatically."
+        val emptyId = ""
+
+        val pslSuffix =
+            parsedStructure.webDomain?.let { pslSupport.fromWebDomain(it) }
+            ?: pslSupport.fromPackageName(parsedStructure.packageName)
+
+        pslSuffix.take(1)
+            .map { suffix ->
+                val domain = "https://${suffix.fullDomain}"
+                ServerPassword(
+                    id = emptyId,
+                    hostname = domain,
+                    username = capturedUsername,
+                    password = capturedPassword,
+                    formSubmitURL = parsedStructure.webDomain ?: domain
```

Comment:
Is `parsedStructure.webDomain` a domain or origin? `formSubmitURL` needs to be an origin.